core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -893,50 +893,64 @@ class KafkaApis(val requestChannel: RequestChannel,
-          val topicDirs = new ZKGroupTopicDirs(offsetFetchRequest.groupId, topicPartition.topic)
-          try {
-            if (!metadataCache.contains(topicPartition.topic))
-              (topicPartition, unknownTopicPartitionResponse)
-            else {
-              val payloadOpt = zkUtils.readDataMaybeNull(s"${topicDirs.consumerOffsetDir}/${topicPartition.partition}")._1
-              payloadOpt match {
-                case Some(payload) =>
-                  (topicPartition, new OffsetFetchResponse.PartitionData(payload.toLong, "", Errors.NONE.code))
-                case None =>
-                  (topicPartition, unknownTopicPartitionResponse)
-              }
-            }
-          } catch {
-            case e: Throwable =>
-              (topicPartition, new OffsetFetchResponse.PartitionData(OffsetFetchResponse.INVALID_OFFSET, "",
-                Errors.forException(e).code))
-          }
-        }.toMap
-        new OffsetFetchResponse((responseInfo ++ unauthorizedStatus).asJava)
+      // reject the request if not authorized to the group
+      if (!authorize(request.session, Read, new Resource(Group, offsetFetchRequest.groupId))) {
+        if (header.apiVersion >= 2)
+          new OffsetFetchResponse(Errors.GROUP_AUTHORIZATION_FAILED.code)
+        else {
+          val unauthorizedGroupResponse = new OffsetFetchResponse.PartitionData(OffsetFetchResponse.INVALID_OFFSET, "", Errors.GROUP_AUTHORIZATION_FAILED.code)
+          val results = offsetFetchRequest.partitions.asScala.map { topicPartition => (topicPartition, unauthorizedGroupResponse)}.toMap
+          new OffsetFetchResponse(results.asJava, header.apiVersion)
+        }
       } else {
-        // version 1 reads offsets from Kafka;
-        val offsets = coordinator.handleFetchOffsets(offsetFetchRequest.groupId, authorizedTopicPartitions).toMap
+        val partitions =
+          if (offsetFetchRequest.isAllPartitions()) // header.apiVersion == 2
+            coordinator.getPartitions(offsetFetchRequest.groupId)
+          else
+            offsetFetchRequest.partitions.asScala.toList
 
-        // Note that we do not need to filter the partitions in the
-        // metadata cache as the topic partitions will be filtered
-        // in coordinator's offset manager through the offset cache
-        new OffsetFetchResponse((offsets ++ unauthorizedStatus).asJava)
-      }
+        val (authorizedTopicPartitions, unauthorizedTopicPartitions) = partitions.partition { topicPartition =>
+          authorize(request.session, Describe, new Resource(auth.Topic, topicPartition.topic))
+        }
+
+        val unknownTopicPartitionResponse = new OffsetFetchResponse.PartitionData(OffsetFetchResponse.INVALID_OFFSET, "", Errors.UNKNOWN_TOPIC_OR_PARTITION.code)
+        val unauthorizedStatus = unauthorizedTopicPartitions.map(topicPartition => (topicPartition, unknownTopicPartitionResponse)).toMap
```

Comment:
I think you need to be careful about listing unauthorized topics. If `offsetFetchRequest.isAllPartitions()` is `true`, then you shouldn't reveal the existence of unauthorized topics. See `handleTopicMetadataRequest` for an example of what I mean.